src/metabase/query_processor/parameters/dates.clj

```diff
@@ -0,0 +1,580 @@
+(ns metabase.query-processor.parameters.dates
+  "Shared code for handling datetime parameters, used by both MBQL and native params implementations."
+  (:require
+   [clojure.string :as str]
+   [java-time.api :as t]
+   [medley.core :as m]
+   [metabase.lib.core :as lib]
+   [metabase.lib.schema.common :as lib.schema.common]
+   [metabase.lib.schema.expression :as lib.schema.expression]
+   [metabase.lib.schema.id :as lib.schema.id]
+   [metabase.lib.schema.parameter :as lib.schema.parameter]
+   [metabase.query-processor.error-type :as qp.error-type]
+   [metabase.query-processor.timezone :as qp.timezone]
+   [metabase.util.date-2 :as u.date]
+   [metabase.util.i18n :refer [tru]]
+   [metabase.util.malli :as mu]
+   [metabase.util.malli.registry :as mr]
+   [metabase.util.time :as u.time])
+  (:import
+   (java.time.temporal Temporal)))
+
+(set! *warn-on-reflection* true)
+
+(def ^:private temporal-units-regex #"(millisecond|second|minute|hour|day|week|month|quarter|year)")
```

Comment:
Is it intentional that millisecond is allowed here but later (e.g., in `operations-by-date-unit`) it's missing?